packages/uploadthing/test/__test-helpers.ts

```diff
@@ -94,111 +77,233 @@ const mockPresigned = (file: {
   };
 };
 
-export const mockExternalRequests =
-  (db: MockDbInterface): FetchEsque =>
-  async (_url, init) => {
-    fetchMock(_url, init);
-    if (_url instanceof Request) return new Response("Wut?", { status: 500 });
-    const url = new URL(_url);
-
-    // Pass through localhost requests
-    if (url.host.startsWith("localhost")) {
-      return fetch(_url, init);
-    }
-
-    // Mock file downloads
-    if (url.host === "cdn.foo.com") {
-      return new Response("Lorem ipsum doler sit amet", {
-        headers: { "Content-Type": "text/plain" },
-      });
-    }
-
-    // Mock UT Api
-    if (url.host === "uploadthing.com") {
-      const body = (() => {
-        if ((init?.method ?? "GET") === "GET") return null;
-        const json = JSON.parse(init?.body as string);
-        if (!isObject(json)) throw new Error("Expected body to be an object");
-        return json;
-      })();
-
-      if (url.pathname === "/api/uploadFiles") {
-        const presigneds = (body?.files as any[]).map((file) => {
+export const msw = setupServer(
+  /**
+   * S3
+   */
+  http.post("https://bucket.s3.amazonaws.com", async ({ request }) => {
+    s3Mock(request.url, {
+      method: request.method,
+      headers: Object.fromEntries(request.headers.entries()),
+      body: await request.formData(),
+    });
+    return HttpResponse.json(null, { status: 200 });
+  }),
+  http.put("https://bucket.s3.amazonaws.com/:key", async ({ request }) => {
+    s3Mock(request.url, {
+      method: request.method,
+      headers: Object.fromEntries(request.headers.entries()),
+      body: await request.blob(),
+    });
+    return HttpResponse.json(null, {
+      status: 204,
+      headers: { ETag: "abc123" },
+    });
+  }),
+  /**
+   * Static Assets
+   */
+  http.get("https://cdn.foo.com/:fileKey", ({ request }) => {
+    staticAssetMock(request.url);
+
+    return HttpResponse.text("Lorem ipsum doler sit amet", {
+      headers: { "content-type": "text/plain" },
+    });
+  }),
+  http.get("https://utfs.io/f/:key", ({ request }) => {
+    staticAssetMock(request.url);
+
+    return HttpResponse.text("Lorem ipsum doler sit amet", {
+      headers: { "content-type": "text/plain" },
+    });
+  }),
+);
+beforeAll(() => msw.listen({ onUnhandledRequest: "bypass" }));
+afterAll(() => msw.close());
+
+export interface MockDbInterface {
+  files: any[];
+  insertFile: (file: any) => void;
+  getFileByKey: (key: string) => any;
+}
+
+/**
+ * Prepend MSW listeners to mock the UploadThing API
+ * Provide the `db` instance to store data within the test
+ */
+export const useDb = (db: MockDbInterface) =>
+  msw.use(
+    http.post(
+      "https://uploadthing.com/api/prepareUpload",
+      async ({ request }) => {
+        const body = (await request.json()) as { files: any[] } & Record<
```

Comment:
You can annotate the request body if you are reading it as json using the generic type argument to the `http.post()` handler:

```ts
http.post<{ files: File[] }>("https://uploadthing.com/api/prepareUpload", async ({ request }) => {
  const body = await request.json() 
  body // { files: File[] }
})
```

The advantage of this to keep your resolvers tidier.